### DISCUSSION TITLE
Install workspace root package in itself

### DISCUSSION THREAD

**Discussion by hubertkuoch** [pnpm/pnpm#5453] [Q&A]
Is there any downside in installing a `pnpm` workspace root package in itself? The reason of doing this is essentially when in dev mode, e.g. when we have some bin used across all packages, or if say we want to share tsconfig (AFAIK referencing package by its name works for JS files but not for JSON) without having to create separate dedicated packages.

I tried it and it works well but I get the below (workspace is the root package):

<img width="486" alt="Screenshot 2022-10-05 at 09 35 02" src="https://user-images.githubusercontent.com/50402245/194006735-07df158e-a93f-427d-a4de-036359154ba0.png">

Just wondering whether the `???` means there is anything to consider that I may have missed?

**Comment by hubertkuoch** [comment]
I have done so and have not encountered any issue so far by doing this.

This allows me to intentionally publicly hoist all tooling related dev dependencies e.g. CLIs, reusable tsconfig, etc. to the workspace root and hence have them all available for all packages without having to bloat .npmrc file if I were to do the same with dedicated local packages.

@zkochan Happy to get your view since I get inspired by the pnpm repo and noticed you do install each package as a dev dependency of itself, just wanted to make sure there is no any counter indication in doing this for workspace root as well.

**Reply by hubertkuoch** [reply]
Just to complement with a concrete use case of why I am doing this, in monorepo we usually create new branch by checking out with git sparse index and hence there is an internal script in charge to clean all cached artifacts including node_modules.

If the scripts are encapsulated in a dedicated local package and installed in the root as an executable bin, the root clean task will fail if the former gets cleaned before it and same the other way around. Hence one possibility is to have all the scripts directly at the root level. Installing the root as a dev dependency of itself is the only solution I find to make the scripts available in all the workspace subdir packages.

**Reply by zkochan** [reply]
I guess it should be OK.

But using hoisting I don't recommend. Hoisting is only for tools that don't work otherwise. Don't use it as a feature.

**Reply by hubertkuoch** [reply]
Thanks @zkochan for your answer, highly appreciated.

Sorry, just to make sure I get you mean, say I have a reusable *clean* script that uses an external dependency (e.g. `rimraf`) to remove node_modules, and say this script is needed in all packages including `root`:

- option 1: create `clean` package (with `rimraf` in its package.json) and install it in `root`. In that case, all packages have access to it (incl. `root`). The problem arises if `clean` gets *cleaned* before `root`, the same task will fail in `root` because `rimraf` will have been removed. So the workaround in that case is to either install `rimraf` also in `root` OR hoist it from `clean` to `root` via .npmrc `public-hoist-pattern` (in that case, no need to add `rimraf` in root).

- option 2 (the one I described above):  don't create `clean` package just for the sake of sharing it, but instead add `rimraf` and define *clean.js* directly in `root` and export it from `root` package.json, that is why installing `root` in itself is needed to make it importable from other packages.

In both cases, the end result is similar (i.e. with `rimraf` *hoisted* in `root` so to speak) but I think option 2 is better because all shared tools are defined in the same place and explicitly stated as available throughout the workspace. Do you also consider option 2 as being a non-recommended use of hoisting as a feature?
